include/pybind11/detail/function_record.h

```diff
@@ -0,0 +1,674 @@
+#pragma once
+
+#include "common.h"
+#include "internal_pytypes.h"
+#include "python_compat.h"
+#include "vendor/optional.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename... Args>
+struct process_attributes;
+
+template <typename T>
+using is_call_guard = is_instantiation<call_guard, T>;
+
+/// Extract the ``type`` from the first `call_guard` in `Extras...` (or `void_type` if none
+/// found)
+template <typename... Extra>
+using extract_guard_t = typename exactly_one_t<is_call_guard, call_guard<>, Extra...>::type;
+
+// A version of min that is usable in constexpr contexts
+template <typename T>
+constexpr T constexpr_min(T a, T b) {
+    return a > b ? b : a;
+}
+
+/// Check the number of named arguments at compile time
+template <typename... Extra,
+          size_t named = constexpr_sum(std::is_base_of<arg, Extra>::value...),
+          size_t self = constexpr_sum(std::is_same<is_method, Extra>::value...)>
+constexpr bool expected_num_args(size_t nargs, bool has_args, bool has_kwargs) {
+    PYBIND11_WORKAROUND_INCORRECT_MSVC_C4100(nargs, has_args, has_kwargs);
+    return named == 0 || (self + named + size_t(has_args) + size_t(has_kwargs)) == nargs;
+}
+
+/// Internal data structure which holds metadata about a keyword argument
+struct argument_record {
+    std::string name;
+    std::string desc;
+    object value;     ///< Associated Python object
+    bool convert : 1; ///< True if the argument is allowed to convert when loading
+    bool none : 1;    ///< True if None is allowed when loading
+
+    argument_record() : name(""), desc(""), value(), convert(true), none(true) {}
```

Comment:
```suggestion
    std::string name = "";
    std::string desc = "";
    object value();     ///< Associated Python object
    bool convert : 1 = true; ///< True if the argument is allowed to convert when loading
    bool none : 1 = true;    ///< True if None is allowed when loading

    argument_record() = default;
```

Reply:
The bitfields mean that this doesn't work.